High CPU Load

Hello,

our Monitoring + htop shows extremely High CPU usage when using the latest standalone.

![grafik](https://github.com/user-attachments/assets/11ee66e4-c76f-4ba8-b2d2-1dd02342747f)

The drop off u see here was caused by a restart of the standalone.

Java Version:
```
openjdk version "17.0.12" 2024-07-16
OpenJDK Runtime Environment (build 17.0.12+7-Ubuntu-1ubuntu222.04)
OpenJDK 64-Bit Server VM (build 17.0.12+7-Ubuntu-1ubuntu222.04, mixed mode, sharing)
```

Used Standalone:
- https://github.com/MCXboxBroadcast/Broadcaster/releases/download/58/MCXboxBroadcastStandalone.jar

[18:56:20 INFO] [geyser]: [mcxboxbroadcast] [Primary Session] Successfully authenticated as Utopiverse (2535421049843580)
>....[27821.487s][warning][os,thread] Failed to start thread "Unknown thread" - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[27821.487s][warning][os,thread] Failed to start the native thread for java.lang.Thread "HttpClient-95-Worker-1"
[18:56:20 ERROR] [geyser]: [mcxboxbroadcast] [Primary Session] Failed to check profile settings
java.io.IOException: unable to create native thread: possibly out of memory or process/resource limits reached
        at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:966) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:133) ~[java.net.http:?]
        at com.github.mizosoft.methanol.Methanol$InterceptorChain.forward(Methanol.java:828) ~[?:?]
        at com.github.mizosoft.methanol.Methanol$AutoDecompressingInterceptor.intercept(Methanol.java:937) ~[?:?]
        at com.github.mizosoft.methanol.Methanol$InterceptorChain.forward(Methanol.java:833) ~[?:?]
        at com.github.mizosoft.methanol.Methanol$RequestRewritingInterceptor.intercept(Methanol.java:881) ~[?:?]
        at com.github.mizosoft.methanol.Methanol$InterceptorChain.forward(Methanol.java:833) ~[?:?]
        at com.github.mizosoft.methanol.Methanol.send(Methanol.java:316) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.SessionManagerCore.checkGamertagUpdate(SessionManagerCore.java:492) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.SessionManagerCore.init(SessionManagerCore.java:148) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.SessionManager.init(SessionManager.java:79) ~[?:?]
        at com.rtm516.mcxboxbroadcast.bootstrap.geyser.MCXboxBroadcastExtension.createSession(MCXboxBroadcastExtension.java:213) ~[?:?]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached
        at java.base/java.lang.Thread.start0(Native Method) ~[?:?]
        at java.base/java.lang.Thread.start(Thread.java:1553) [?:?]
        at java.base/java.lang.System$2.start(System.java:2577) ~[?:?]
        at java.base/jdk.internal.vm.SharedThreadContainer.start(SharedThreadContainer.java:152) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:953) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1375) ~[?:?]
        at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.ensureExecutedAsync(HttpClientImpl.java:183) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.execute(HttpClientImpl.java:175) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.SocketTube$InternalWriteSubscriber.startSubscription(SocketTube.java:396) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.AsyncTriggerEvent.handle(AsyncTriggerEvent.java:54) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.run(HttpClientImpl.java:1314) ~[java.net.http:?]
>....[27821.488s][warning][os,thread] Failed to start thread "Unknown thread" - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[18:56:20 INFO] [geyser]: [mcxboxbroadcast] [Primary Session] Creating Xbox LIVE session...
>....[27821.488s][warning][os,thread] Failed to start the native thread for java.lang.Thread "HttpClient-95-Worker-2"
[27821.570s][warning][os,thread] Failed to start thread "Unknown thread" - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[27821.570s][warning][os,thread] Failed to start the native thread for java.lang.Thread "pool-98-thread-2"
[18:56:20 ERROR] [geyser]: [mcxboxbroadcast] [Auth] Failed to get/refresh auth token
java.io.IOException: unable to create native thread: possibly out of memory or process/resource limits reached
        at java.net.http/jdk.internal.net.http.HttpClientImpl.send(HttpClientImpl.java:966) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.HttpClientFacade.send(HttpClientFacade.java:133) ~[java.net.http:?]
        at net.lenni0451.commons.httpclient.executor.HttpClientExecutor.execute(HttpClientExecutor.java:48) ~[?:?]
        at net.lenni0451.commons.httpclient.HttpClient.execute(HttpClient.java:232) ~[?:?]
        at net.lenni0451.commons.httpclient.HttpClient.execute(HttpClient.java:213) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.AuthManager.fetchPlayfabSessionTicket(AuthManager.java:142) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.AuthManager.initialise(AuthManager.java:129) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.AuthManager.getPlayfabSessionTicket(AuthManager.java:169) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.SessionManagerCore.setupSession(SessionManagerCore.java:375) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.SessionManagerCore.createSession(SessionManagerCore.java:196) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.SessionManagerCore.init(SessionManagerCore.java:162) ~[?:?]
        at com.rtm516.mcxboxbroadcast.core.SessionManager.init(SessionManager.java:79) ~[?:?]
        at com.rtm516.mcxboxbroadcast.bootstrap.geyser.MCXboxBroadcastExtension.createSession(MCXboxBroadcastExtension.java:213) ~[?:?]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached
        at java.base/java.lang.Thread.start0(Native Method) ~[?:?]
        at java.base/java.lang.Thread.start(Thread.java:1553) [?:?]
        at java.base/java.lang.System$2.start(System.java:2577) ~[?:?]
        at java.base/jdk.internal.vm.SharedThreadContainer.start(SharedThreadContainer.java:152) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:953) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1375) ~[?:?]
        at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.ensureExecutedAsync(HttpClientImpl.java:183) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.execute(HttpClientImpl.java:175) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.SocketTube$InternalWriteSubscriber.startSubscription(SocketTube.java:396) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.AsyncTriggerEvent.handle(AsyncTriggerEvent.java:54) ~[java.net.http:?]
        at java.net.http/jdk.internal.net.http.HttpClientImpl$SelectorManager.run(HttpClientImpl.java:1314) ~[java.net.http:?]
[18:56:26 INFO] [geyser]: [mcxboxbroadcast] [Primary Session] Starting SessionManager...
[18:56:27 INFO] [geyser]: [mcxboxbroadcast] [Primary Session] Successfully authenticated as Utopiverse (2535421049843580)
[18:56:28 INFO] [geyser]: [mcxboxbroadcast] [Primary Session] Creating Xbox LIVE session...
>....[27829.615s][warning][os,thread] Failed to start thread "Unknown thread" - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[27829.615s][warning][os,thread] Failed to start the native thread for java.lang.Thread "pool-102-thread-1"
[27829.619s][warning][os,thread] Failed to start thread "Unknown thread" - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
[27829.620s][warning][os,thread] Failed to start the native thread for java.lang.Thread "Thread-116"
[18:56:28 ERROR]: Exception in thread "WebSocketConnectReadThread-3313" java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached
[18:56:28 ERROR]:       at java.base/java.lang.Thread.start0(Native Method)
[18:56:28 ERROR]:       at java.base/java.lang.Thread.start(Thread.java:1526)
[18:56:28 ERROR]:       at org.java_websocket.client.WebSocketClient.run(WebSocketClient.java:499)
[18:56:28 ERROR]:       at java.base/java.lang.Thread.run(Thread.java:1583)

--- a/core/src/main/java/com/rtm516/mcxboxbroadcast/core/RtaWebsocketClient.java
+++ b/core/src/main/java/com/rtm516/mcxboxbroadcast/core/RtaWebsocketClient.java
@@ -5,6 +5,7 @@ import org.java_websocket.client.WebSocketClient;
 import org.java_websocket.handshake.ServerHandshake;
 
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
 
 /**
  * Handle the connection and authentication with the RTA websocket
@@ -16,6 +17,7 @@ public class RtaWebsocketClient extends WebSocketClient {
     private final Logger logger;
     private final String xuid;
     private boolean firstConnectionId = true;
+    private CompletableFuture<String> connectionIdFuture = new CompletableFuture<>();
 
     /**
      * Create a new websocket and add the Authorization header
@@ -37,6 +39,10 @@ public class RtaWebsocketClient extends WebSocketClient {
      */
     public String getConnectionId() {
         return connectionId;
+    }
+
+    public CompletableFuture<String> getConnectionIdFuture() {
+        return connectionIdFuture;
     }
 
     /**
@@ -64,6 +70,7 @@ public class RtaWebsocketClient extends WebSocketClient {
         if (message.contains("ConnectionId") && firstConnectionId) {
             Object[] parts = Constants.GSON.fromJson(message, Object[].class);
             connectionId = ((Map<String, String>) parts[4]).get("ConnectionId");
+            connectionIdFuture.complete(connectionId);
             firstConnectionId = false;
 
             // Let xbox know we want friend updates

--- a/core/src/main/java/com/rtm516/mcxboxbroadcast/core/SessionManagerCore.java
+++ b/core/src/main/java/com/rtm516/mcxboxbroadcast/core/SessionManagerCore.java
@@ -6,17 +6,15 @@ import com.rtm516.mcxboxbroadcast.core.exceptions.SessionCreationException;
 import com.rtm516.mcxboxbroadcast.core.exceptions.SessionUpdateException;
 import com.rtm516.mcxboxbroadcast.core.models.auth.SessionStartBody;
 import com.rtm516.mcxboxbroadcast.core.models.auth.SessionStartResponse;
+import com.rtm516.mcxboxbroadcast.core.models.auth.XboxTokenInfo;
 import com.rtm516.mcxboxbroadcast.core.models.other.ProfileSettingsResponse;
 import com.rtm516.mcxboxbroadcast.core.models.session.CreateHandleRequest;
 import com.rtm516.mcxboxbroadcast.core.models.session.CreateHandleResponse;
 import com.rtm516.mcxboxbroadcast.core.models.session.SessionRef;
 import com.rtm516.mcxboxbroadcast.core.models.session.SocialSummaryResponse;
-import com.rtm516.mcxboxbroadcast.core.models.auth.XboxTokenInfo;
 import com.rtm516.mcxboxbroadcast.core.notifications.NotificationManager;
 import com.rtm516.mcxboxbroadcast.core.storage.StorageManager;
-
 import com.rtm516.mcxboxbroadcast.core.webrtc.RtcWebsocketClient;
-import org.java_websocket.enums.ReadyState;
 
 import java.io.IOException;
 import java.net.URI;
@@ -24,9 +22,7 @@ import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse;
 import java.time.Duration;
-import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
@@ -342,33 +338,11 @@ public abstract class SessionManagerCore {
      * @return The received connection ID
      */
     protected Future<String> waitForConnectionId() {
-        CompletableFuture<String> completableFuture = new CompletableFuture<>();
-
-        Executors.newCachedThreadPool().submit(() -> {
-            while (rtaWebsocket.getConnectionId() == null) {
-                Thread.sleep(100);
-            }
-            completableFuture.complete(rtaWebsocket.getConnectionId());
-
-            return null;
-        });
-
-        return completableFuture;
+        return this.rtaWebsocket.getConnectionIdFuture();
     }
 
     protected Future<Void> waitForRTCConnection() {
-        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
-
-        Executors.newCachedThreadPool().submit(() -> {
-            while (rtcWebsocket.getReadyState() == ReadyState.NOT_YET_CONNECTED) {
-                Thread.sleep(100);
-            }
-            completableFuture.complete(null);
-
-            return null;
-        });
-
-        return completableFuture;
+        return this.rtcWebsocket.onOpenFuture();
     }
 
     protected String setupSession() {

--- a/core/src/main/java/com/rtm516/mcxboxbroadcast/core/webrtc/RtcWebsocketClient.java
+++ b/core/src/main/java/com/rtm516/mcxboxbroadcast/core/webrtc/RtcWebsocketClient.java
@@ -17,6 +17,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
@@ -45,6 +46,7 @@ public class RtcWebsocketClient extends WebSocketClient {
     private final List<CandidateHarvester> candidateHarvesters;
 
     private ScheduledFuture<?> heartbeatFuture;
+    private CompletableFuture<Void> onOpenFuture = new CompletableFuture<>();
 
     /**
      * Create a new websocket and add the Authorization header
@@ -69,6 +71,10 @@ public class RtcWebsocketClient extends WebSocketClient {
         this.candidateHarvesters = new ArrayList<>();
     }
 
+    public CompletableFuture<Void> onOpenFuture() {
+        return onOpenFuture;
+    }
+
     /**
      * When the web socket connects start the heartbeat to keep the connection alive
      *
@@ -82,6 +88,7 @@ public class RtcWebsocketClient extends WebSocketClient {
         heartbeatFuture = scheduledExecutorService.scheduleWithFixedDelay(() -> {
             send(Constants.GSON.toJson(new WsToMessage(0, null, null)));
         }, 40, 40, TimeUnit.SECONDS);
+        onOpenFuture.complete(null);
     }
 
     /**

